### ISSUE TITLE
ENTRYPOINT not called on the container when launched using this plugin

### ISSUE THREAD

**Issue by CJCombrink** [jenkinsci/docker-swarm-plugin#53]
My (Windows) container needs to call a specific script when starting up, this script is defined in my Dockerfile as:
```
ENTRYPOINT C:\entrypoint.bat &&
```

For some reason this script is not called when the Docker Swarm plugin runs the container in my swarm. The container is started correctly but my pipeline fails since some functionality is missing since the entrypoint script did not run.

If I replace the "sh" part of the default "Command" (Docker agent template configuration) with:
```
cmd /c C:\entrypoint.bat && sh
```
Then the container is started correctly, the entry script is run, and the node connects correctly and finally my pipeline succeeds. 

Is there perhaps a reason why the ENTRYPOINT is not called by the plugin?

Although I have a workaround in place, any help will be appreciated to prevent the need for the workaround. 

PS: The entrypoint script runs the vcvars64.bat file to set up the Visual Studio environment inside the container for building to take place.

**Comment by Roemer** [comment]
Just one question in front: Did you check the `Windows Container` flag in the placement? If not, are you manually adding the agent / the agent connection parameters into the container?

**Comment by CJCombrink** [comment]
> Did you check the Windows Container flag in the placement?

Yes I did and it made no difference, I reverted to uncheck the flag since my container is set up to correctly have sh and java available so that I can run the exact same pipelines across windows and Linux. I reverted the `Windows Container flag` since I was not sure how it would affect the addition of my command added before 'sh' (The help states that it will replace sh with PowerShell alternatives and I was not clear how it would do so to also allow me to call my script, the word 'overrides' made me cautious)

I have not changed anything from the default parameters except adding to run my script as a workaround, my original command was still:
```
sh
-cx
curl --connect-timeout 20 --max-time 60 -o agent.jar $DOCKER_SWARM_PLUGIN_JENKINS_AGENT_JAR_URL && java -jar agent.jar -jnlpUrl $DOCKER_SWARM_PLUGIN_JENKINS_AGENT_JNLP_URL -secret $DOCKER_SWARM_PLUGIN_JENKINS_AGENT_SECRET -noReconnect -workDir /tmp
```
With `Working Directory` set to `C:\Jenkins`

Thus now it is:
```
cmd /c C:\entrypoint.bat && sh
-cx
curl --connect-timeout 20 --max-time 60 -o agent.jar $DOCKER_SWARM_PLUGIN_JENKINS_AGENT_JAR_URL && java -jar agent.jar -jnlpUrl $DOCKER_SWARM_PLUGIN_JENKINS_AGENT_JNLP_URL -secret $DOCKER_SWARM_PLUGIN_JENKINS_AGENT_SECRET -noReconnect -workDir /tmp
```

**Comment by Roemer** [comment]
About the Windows Container flag: Yes, currently if it is set it completely overrides the command by a hardcoded powershell version (see https://github.com/jenkinsci/docker-swarm-plugin/blob/4ecfe7e7acf6ff0b134983a628e6b6b1c02dbfc0/src/main/java/org/jenkinsci/plugins/docker/swarm/DockerSwarmComputerLauncher.java#L88). That is marked as todo to have a better alternative (maybe just remove the flag and show the powershell and sh version in the help).

I actually do not know how the entrypoint behaves when a command is also given when spinning up a Docker Swarm Service. Possibly the command does overwrite the entrypoint entirely on Docker side? We might need to verify the behavior in order to understand if the workaround is actually the correct way or not and something is not working correctly.

**Comment by CJCombrink** [comment]
Unfortunately I have no knowledge of how the Docker Swarm API works and handles the ENTRYPOINT and the normal CMD command, thus I can not comment on if the issue is with this plugin or with the Docker Swarm itself.
